core/src/blocktree.rs

```diff
@@ -340,11 +351,15 @@ impl Blocktree {
         // Handle chaining for the working set
         handle_chaining(&db, &mut write_batch, &slot_meta_working_set)?;
         let mut should_signal = false;
+        let mut newly_completed_slots = vec![];
 
         // Check if any metadata was changed, if so, insert the new version of the
         // metadata into the write batch
         for (slot, (meta, meta_backup)) in slot_meta_working_set.iter() {
             let meta: &SlotMeta = &RefCell::borrow(&*meta);
+            if self.completed_slots.len() > 0 && is_newly_completed_slot(meta, meta_backup) {
```

Comment:
is the first check an optimization?

Reply:
nevermind...

this code would be more readable if you renamed completed_slots to completed_slots_senders